repositories/heartbeart.go

```diff
@@ -77,6 +78,19 @@ func (r *HeartbeatRepository) GetAllWithin(from, to time.Time, user *models.User
 	return heartbeats, nil
 }
 
+func (r *HeartbeatRepository) GetAllWithinUTC(from, to time.Time, user *models.User) ([]*models.Heartbeat, error) {
```

Comment:
A few thoughts about this. First, what was your intention with this method? Are you sure this actually does, what you want? In my understanding, this might lead to errors. At least in the case of MySQL, the database connection has a timezone set, which is by default the host's timezone. That is, all dates will be interpreted as that timezone. If you now convert to UTC, won't there be a wrong offset. I once did a bit of research on that topic and took a couple of notes [here](https://github.com/muety/wakapi/blob/7159df30c25a882434185395c7f4927bad34f110/config/db.go#L12), maybe that helps. Would be nice if you quickly shared your thoughts on this timezone conversion.

In any case, I'd rather like the method to be as generic as possible, i.e. `GetAllWithinTZ()` and then pass a timezone as another parameter. Just more flexible.

Reply:
According to wakatime API docs, `date` is "Requested day; Heartbeats will be returned from 12am until 11:59pm in user's timezone for this day." 
The intention is to take the date, and query the database for heartbeats between the start and end of the specified date in the user's timezone (from user settings). 
I added this function because using the other function doesn't seem to produce the correct results. 

Some preliminary investigation:
GetAllWithin: `time >= "2022-01-02 00:00:00" AND time < "2022-01-03 00:00:00"`
GetAllWithinUTC: `time >= "2022-01-01 13:00:00" AND time < "2022-01-02 13:00:00"`

Running `sqlite3` manually, 
```
select * from heartbeats WHERE time >= "2022-01-01 13:00:00" AND time < "2022-01-02 13:00:00";
first entry timestamps are 2022-01-01 23:10:38.963+00:00, 2022-01-01 23:10:19.221+00:00
last entry timestamps are 2022-01-02 12:01:42.779+00:00, 2022-01-02 12:01:23.354+00:00

select * from heartbeats WHERE time >= "2022-01-02 00:00:00" AND time < "2022-01-03 00:00:00";
first entry timestamps are 2022-01-02 00:06:05.196+00:00, 2022-01-02 00:05:45.332+00:00
last entry timestamps are 2022-01-02 23:59:43.287+00:00, 2022-01-02 23:59:23.73+00:00
```
When you consider that my timezone is +11, the first appears to be correct. 
However, I do understand the concern around this. 
I'll take a look at your notes and investigate further.

Reply:
```go
package main;
import (
    "time"
    "fmt"
)
func main() {
    // User location
    tz, _ := time.LoadLocation("Australia/Melbourne")
    // The input date
    in, _ := time.Parse("2006-01-02", "2022-01-01")

    // utils.StartOfDay(in.In(tz))
    l := in.In(tz)
    d := time.Date(l.Year(), l.Month(), l.Day(), 0, 0, 0, 0, l.Location());

    // GetAllWithinUTC (old), GetAllWithinUnix (new implementation)
    fmt.Println(d.UTC())
    fmt.Println(d.Unix())
    fmt.Println(d.UTC().Unix() == d.Unix())

    // GetAllWithin
    fmt.Println(d.Local())
}
```
Output
```
2021-12-31 13:00:00 +0000 UTC
1640955600
true

2022-01-01 00:00:00 +1100 AEDT
```

An example to show the difference between the `GetAllWithin` and the new implementation.

Reply:
`DATETIME(..., 'unixepoch')` doesn't seem to be a valid MySQL function.

```
2022-02-05T15:06:15.511502714+01:00 [ERROR] failed to retrieve heartbeats - Error 1064: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '(?, 'unixepoch') AND time < datetime(?, 'unixepoch') ORDER BY time asc' at line 1
```

---

Normally, we shouldn't even have to care about converting timezones back and forth, that's something that the driver should handle for you.

---

I'm pretty sure you can just use `GetAllWithin()` as is. Here's an example.
* Using SQLite for now
* I'm in `Euopre/Berlin` (UTC+1)
* E.g. `2022-01-05 23:59` in my time zone corresponds to `2022-01-05 22:59` in UTC
* I have only one heartbeat in my database at exactly that time: 
![image](https://user-images.githubusercontent.com/8978843/152646488-62848bb3-a2b3-40aa-989c-6c3f33e4f749.png)
* I modified your endpoint to use `GetAllWithin()` instead of `GetAllWithinUnix()`
* I call `/api/compat/wakatime/v1/users/current/heartbeats?date=2022-02-05`
* According to WakaTime docs, this request shall return all heartbeats that were created on Feb 5th  from my point of view, i.e. in my timezone. Accordingly, I would expect the above heartbeat to be returned. 
* And it actually is:
![image](https://user-images.githubusercontent.com/8978843/152646622-eebddea2-4012-416b-a054-bb2e05e450b8.png)
* The server log reads: ```SELECT * FROM `heartbeats` WHERE `heartbeats`.`user_id` = "n1try" AND time >= "2022-02-05 00:00:00" AND time < "2022-02-06 00:00:00" ORDER BY time asc```
* All fine, in my understanding

---

I think the key to this discussion is, that `time` column is of type `TIMESTAMP`, not `DATETIME`. Timestamp type doesn't store any time zone information, it's neither assumed to be UTC nor anything else. It's just the raw date string and is interpreted as whatever the application wants it to interpret as. My Wakapi server, running in `Europe/Berlin` time zone, retrieves `2022-02-05 23:59:00.000` from the database and interprets it as `2022-02-05 23:59:00.000+01:00`. If I sent you my database, everything would be entirely messed up, because although the data doesn't change at all, your server would interpret that date as `2022-02-05 23:59:00.000+11:00` (which is `2022-02-05 12:59:00.000+00:00`). At least in the case of MySQL, I would assume SQLite and Postgres to work similarly.

Analogously for queries. Since the connection's timezone is, in my case, implicitly set to `Europe/Berlin`, a query `... WHERE time >= 2022-02-05 00:00:00 ...` is implicitly actually `... WHERE time >= 2022-02-05 00:00:00+01:00 ...` then, so to say, although this "conversion" never explicitly happens.

Although I'm quite confident with this, I might still be wrong. But then, please prep a very simple example SQLite database that I can use to prove so. Try to avoid issuing raw queries, but test everything "end to end", i.e. insert and query through Wakapi API.

Reply:
> DATETIME(..., 'unixepoch') doesn't seem to be a valid MySQL function.

Thanks for the catch, I didn't consider database differences. 

> I'm pretty sure you can just use GetAllWithin() as is. 

Yes, you're right. 
I reverted the change and modified the postman tests to hardcoded dates to specifically test for this and it worked end to end. 

> If I sent you my database, everything would be entirely messed up. 

I will probably need to do more experimentation to fully understand this. 
However, I feel that this behavior is not ideal. 
Perhaps we can create an issue for further investigation/fixes?